### PR TITLE
Feature/c 867 organization enrichment

### PR DESCRIPTION
**Pull request by elayira**
# Changes proposed ✍️

### <samp>🤖 Generated by Copilot at 5415868</samp>

This pull request implements the organization enrichment feature, which allows enriching the organization data of the users using the People Data Labs (PDL) API. It adds and updates the configuration, models, migrations, repositories, feature flags, message types, and services for the feature. It also adds a new worker function for the bulk organization enrichment operation, which is triggered by a service message of type `OrganizationBulkEnrichMessage`.
​

### <samp>🤖 Generated by Copilot at 5415868</samp>

> _Sing, O Muse, of the mighty pull request_
> _That enriched the data of organizations_
> _With the secret key of PDL, the best_
> _Of all the APIs of the nations._

### <samp>🤖 Generated by Copilot at 5415868</samp>

*  Add a new feature for enriching organization information using the People Data Labs (PDL) API ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-507744d9a63039e8740974e9eb54b7bef83370f87aeb6c5f49c50e3e3c658a54R184-R187), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-f5e4821f147eb37a6ea650b940e2def3d833842f72b97c266aad5d2478d77225R32), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-f5e4821f147eb37a6ea650b940e2def3d833842f72b97c266aad5d2478d77225R228-R233), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-5fcc50392ce9c4b902e2ab9445dcaf1f9f3fff2d8e94244fef9111ecfcf4243eR17), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-5343d3507dbec3f8eea6c3604925d8343c6747aff9265316721c9ae5eaf9fee9R75-R79), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-b8b65cf54228d58772f94e34b19582d291f4fb99d8299147f110d78e7aacca71L1-R202), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-174a13e1169ad442dbb819e8f5a416bdf7f12ba9896145fd4bd1b665e95ed8bbR1-R30), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-732905b9cf71f39b9529ffacfff3870914df8ecbcd2deac348c496e41128cef0R27), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-732905b9cf71f39b9529ffacfff3870914df8ecbcd2deac348c496e41128cef0R33))
  * Define a new interface `OrganizationEnrichmentConfiguration` for the configuration object for the feature in `configTypes.ts` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-507744d9a63039e8740974e9eb54b7bef83370f87aeb6c5f49c50e3e3c658a54R184-R187))
  * Export the `ORGANIZATION_ENRICHMENT_CONFIG` object from `index.ts` in the config folder, which assigns the `apiKey` property based on the `KUBE_MODE` environment variable ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-f5e4821f147eb37a6ea650b940e2def3d833842f72b97c266aad5d2478d77225R32), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-f5e4821f147eb37a6ea650b940e2def3d833842f72b97c266aad5d2478d77225R228-R233))
  * Define a new property `ORGANIZATION_ENRICHMENT` in the `PLAN_LIMITS` object in `isFeatureEnabled.ts`, which sets the limit of organization enrichment operations per month for each plan ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-5fcc50392ce9c4b902e2ab9445dcaf1f9f3fff2d8e94244fef9111ecfcf4243eR17))
  * Define a new type `OrganizationBulkEnrichMessage` in `messageTypes.ts`, which holds the service and tenant id for the bulk organization enrichment operation ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-5343d3507dbec3f8eea6c3604925d8343c6747aff9265316721c9ae5eaf9fee9R75-R79))
  * Define a new class `OrganizationEnrichmentService` in `organizationEnrichmentService.ts`, which provides the logic for the feature, such as validating the plan limit, fetching the data from the PDL API, and updating the database models ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-b8b65cf54228d58772f94e34b19582d291f4fb99d8299147f110d78e7aacca71L1-R202))
  * Define some types for the feature in `organizationEnrichmentTypes.ts`, such as `OrganizationEnrichmentData`, `OrganizationEnrichmentResponse`, and `OrganizationEnrichmentError` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-174a13e1169ad442dbb819e8f5a416bdf7f12ba9896145fd4bd1b665e95ed8bbR1-R30))
  * Define a new property `ORGANIZATION_ENRICHMENT` in the `FeatureFlag` enum and a new property `ORGANIZATION_ENRICHMENT_COUNT` in the `FeatureFlagRedisKey` enum in `common.ts` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-732905b9cf71f39b9529ffacfff3870914df8ecbcd2deac348c496e41128cef0R27), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-732905b9cf71f39b9529ffacfff3870914df8ecbcd2deac348c496e41128cef0R33))
* Add a new function `BulkorganizationEnrichmentWorker` in `bulkOrganizationEnrichmentWorker.ts`, which performs the bulk organization enrichment operation for a given tenant using the `OrganizationEnrichmentService` class ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-3fba44a40bc51efc29fe0b5100d82d61be4c740afdd2b2cc9df9e237003de361R1-R58))
* Add a new case to the switch statement in the `workerFactory` function in `workerFactory.ts`, which handles the `'enrich-organizations'` service message and calls the `BulkorganizationEnrichmentWorker` function ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-f540afd6079e352be077c96f26fc37c996860a5b2ab1fcc633646ccddc7dd6b8R14), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-f540afd6079e352be077c96f26fc37c996860a5b2ab1fcc633646ccddc7dd6b8R28), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-f540afd6079e352be077c96f26fc37c996860a5b2ab1fcc633646ccddc7dd6b8R72-R75))
* Add some new columns to the `organizations` and `organizationCaches` tables in the database, which store the data obtained from the PDL API, such as `lastEnrichedAt`, `employeeCountByCountry`, `type`, `ticker`, `headline`, `profiles`, `naics`, `industry`, and `founded` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-080d19a3708debd548839e7c13651864edf921a27cad2f343c097f9992dd6d6fL1-R8), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-c00eea2e7c1f282ff8d9d6ef899bb48e7aa41b7e686631730b84f3eca8b49734L1-R8), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-a41275558cd949f17542b90d2752c0f312e3570726147e4784c4909e7c10a3ecL1-R8), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-4c200bab5ba9a18d300ac72e53d18eab5e3a2cf501f21f1cc6e6f6971fd6bb43L1-R8))
  * Add the corresponding migration files to the database migrations folder, which drop some unused columns and add the new columns ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-080d19a3708debd548839e7c13651864edf921a27cad2f343c097f9992dd6d6fL1-R8), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-c00eea2e7c1f282ff8d9d6ef899bb48e7aa41b7e686631730b84f3eca8b49734L1-R8), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-a41275558cd949f17542b90d2752c0f312e3570726147e4784c4909e7c10a3ecL1-R8), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-4c200bab5ba9a18d300ac72e53d18eab5e3a2cf501f21f1cc6e6f6971fd6bb43L1-R8))
  * Add the corresponding properties to the `Organization` and `OrganizationCache` models in the database models folder, which have the appropriate data types and constraints ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-0532bedbaaab427bd1bee309432ab2d6ffae937793159b9d627b69e7feb94ac9R91-R131), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-1c58cd30830b133ff520091363368e9a0b9dfa0a555d436dcce5fcc42734707eR86-R126))
* Add a new static method `bulkUpdate` to the `OrganizationRepository` and `OrganizationCacheRepository` classes in the database repositories folder, which allow updating multiple organization instances in bulk using the `bulkCreate` method with the `updateOnDuplicate` and `returning` options ([link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-19bb7618c04f12142b517c86e4988dc08fbf1727569d49437bdbae190c6462bcR91-R104), [link](https://github.com/CrowdDotDev/crowd.dev/pull/825/files?diff=unified&w=0#diff-a52bcdecd97974bf17c43be28a61e4aeb39bfc9b38ff85ef4f9769279499b7ccR62-R73))

## Checklist ✅
- [x] Label appropriately with `Feature`, `Improvement`, or `Bug`.
- [ ] Add screehshots to the PR description for relevant FE changes
- [ ] New backend functionality has been unit-tested.
- [ ] API documentation has been updated (if necessary) (see [docs on API documentation](https://docs.crowd.dev/docs/updating-api-documentation)).
- [ ] [Quality standards](https://github.com/CrowdDotDev/crowd-github-test-public/blob/main/CONTRIBUTING.md#quality-standards) are met.

### CODE REVIEW

File: backend/src/database/models/organization.ts [hunk 1]
```diff
@@ -88,6 +88,46 @@ export default (sequelize) => {
...
+      },
+      naics: {
+        type: DataTypes.ARRAY(DataTypes.JSONB),
+        allowNull: true,
+      },
+      profiles: {
+        type: DataTypes.ARRAY(DataTypes.TEXT),
+        allowNull: true,
+      },
+      headline: {
```
Comment (joanreyero) [inline backend/src/database/models/organization.ts:111]: What is this? And how is it different from the description? If it's semantically the same (or very similar), we should just use description.

File: backend/src/database/models/organization.ts [hunk 2]
```diff
@@ -88,6 +88,46 @@ export default (sequelize) => {
...
       },
+      founded: {
+        type: DataTypes.INTEGER,
+        allowNull: true,
+      },
+      industry: {
+        type: DataTypes.TEXT,
+        allowNull: true,
+      },
+      size: {
```
Comment (joanreyero) [inline backend/src/database/models/organization.ts:99]: What is this? And how is it different from `employees`? If it's semantically the same (or very similar), we should use employeeCount.

File: backend/src/database/models/organization.ts [hunk 3]
```diff
@@ -88,6 +88,46 @@ export default (sequelize) => {
...
+        type: DataTypes.TEXT,
+        allowNull: true,
+      },
+      naics: {
+        type: DataTypes.ARRAY(DataTypes.JSONB),
+        allowNull: true,
+      },
+      profiles: {
```
Comment (joanreyero) [inline backend/src/database/models/organization.ts:107]: How does this differ from github, linkedin, crunchbase etc?

File: backend/src/database/models/organization.ts (hunk 1 above)
Comment (joanreyero) [reply backend/src/database/models/organization.ts:111]: Could you also add comments for all the new fields that are not obvious? Like the ones where I left a comment, and naics, ticker, type...

File: backend/src/database/repositories/organizationRepository.ts [hunk 4]
```diff
@@ -59,6 +59,22 @@ class OrganizationRepository {
...
 
+  static async bulkUpdate<T extends any[]>(
+    data: T,
+    fields: string[],
+    options: IRepositoryOptions,
+  ): Promise<T> {
+    const isValid = new Set(data.filter((org) => org.id).map((org) => org.id)).size !== data.length
+    if (isValid) return [] as T
+
+    const orgs = await options.database.organization.bulkCreate(data, {
```
Comment (joanreyero) [inline backend/src/database/repositories/organizationRepository.ts:70]: I might be misunderstanding something, but the function name is `bulkUpdate` and the operation here is `bulkCreate`. Is this correct? If so, please add a comment explaining why 🙂

File: backend/src/services/premium/enrichment/organizationEnrichmentService.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+      if(org[field] === null){
+        this.fields.push(field)
+      }
+    }
+    return lodash.pick(org, ['id', 'cachId', ...this.fields])
+  }
+
+  async queryTenancyOrganizations(): Promise<IEnrichableOrganization[]> {
+    const options = await SequelizeRepository.getDefaultIRepositoryOptions()
+    const query = `
```
Comment (joanreyero) [inline backend/src/services/premium/enrichment/organizationEnrichmentService.ts:171]: Same as before, let's keep queries in the repo layer. Feel free to use organizations repo.

File: backend/src/services/premium/enrichment/organizationEnrichmentService.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+      query,
+      {
+        type: QueryTypes.SELECT,
+        replacements: {
+          tenantId: this.tenantId,
+          limit: this.maxOrganizationsLimit,
+        },
+      },
+    )
+    return orgs.map(org => this.selectFieldsForEnrichment(org))
```
Comment (joanreyero) [inline backend/src/services/premium/enrichment/organizationEnrichmentService.ts:214]: From [PDL's docs](https://docs.peopledatalabs.com/docs/company-stats) it looks like we obtain:
- `employee_count` (we should have this as `employees`)
- `location`: it's an object with several interesting fields. I think we should store all of them as a JSONB. Except for the GEO code. Let's make a new column for it.

**Review by joanreyero — CHANGES_REQUESTED** [review]
Great job on this! 💪🏼 It's starting to get there.

I left a few comments of things that should be changed, and some questions 🙂

File: backend/src/database/models/organization.ts (hunk 1 above)
Comment (elayira) [reply backend/src/database/models/organization.ts:115]: The `headline` field according to PDL is a piece of brief information about the company. More detailed info can be found in the company's `summary`. This is what I am using to update our `description` field. `naics` is the company's classification according to the NAIC standard. `ticker` is the company's trading symbol. And `type` is the kind of entity the company is, ex an NGO.

File: backend/src/database/models/organization.ts (hunk 2 above)
Comment (elayira) [reply backend/src/database/models/organization.ts:105]: `employees` is the number of workers in the company. `size` is the well, size of the company that's small, medium, large, etc. Size is represented as a range.

File: backend/src/database/models/organization.ts (hunk 3 above)
Comment (elayira) [reply backend/src/database/models/organization.ts]: The profile holds all the social networks a company has. It is designed to hold multiple links to the same network. It is part of the requirement in the task document.

File: backend/src/database/repositories/organizationRepository.ts (hunk 4 above)
Comment (elayira) [reply backend/src/database/repositories/organizationRepository.ts:122]: `bulkCreate` can be used to create/update bulk. It updates on conflict with primary key or unique field when the `updateOnDuplicate` args is used. Comments have been added to the code to reflect this behavior.

File: backend/src/services/premium/enrichment/organizationEnrichmentService.ts (hunk 5 above)
Comment (elayira) [reply backend/src/services/premium/enrichment/organizationEnrichmentService.ts]: Ditto

File: backend/src/services/premium/enrichment/organizationEnrichmentService.ts (hunk 6 above)
Comment (elayira) [reply backend/src/services/premium/enrichment/organizationEnrichmentService.ts]: We are not currently updating `employees` with `employee_count` from PDL as it's not in the task requirement document.

File: backend/src/bin/jobs/organizationEnricher.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    } as NodeWorkerMessageBase)
+  }
+}
+async function queryEnrichableOrganizations(database: any): Promise<string[]> {
+  const query = `
+    SELECT "id"
+    FROM "tenants"
+    WHERE tenants."plan" IN ('Growth')
+      OR (tenants."isTrialPlan" is true AND tenants."plan" = 'Growth')
+    ;
```
[the remark this answers is not in the document]
Comment (elayira) [reply backend/src/bin/jobs/organizationEnricher.ts]: Done!

**Pull request merged**